### ISSUE TITLE
2.0.2 unittest failures regarding `raft.{bind,self}_addr`

### ISSUE THREAD

**Issue by mbanck** [zalando/patroni#1862]
**Describe the bug**

Running the 2.0.2 unit tests on Debian unstable, I get several failing tests like:

```
tests/test_validator.py:163: AssertionError
_____________________________________________________________________________________________________ TestValidator.test_complete_config _____________________________________________________________________________________________________

self = <tests.test_validator.TestValidator testMethod=test_complete_config>, mock_out = <_io.StringIO object at 0x7f32301dc940>, mock_err = <_io.StringIO object at 0x7f32301dcaf0>

    def test_complete_config(self, mock_out, mock_err):
        schema(config)
        output = mock_out.getvalue()
>       self.assertEqual(['postgresql.bin_dir', 'raft.bind_addr', 'raft.self_addr'], parse_output(output))
E       AssertionError: Lists differ: ['postgresql.bin_dir', 'raft.bind_addr', 'raft.self_addr'] != ['postgresql.bin_dir']
E       
E       First list contains 2 additional elements.
E       First extra element 1:
E       'raft.bind_addr'
E       
E       - ['postgresql.bin_dir', 'raft.bind_addr', 'raft.self_addr']
E       + ['postgresql.bin_dir']

tests/test_validator.py:152: AssertionError
```

This also happens when I add  `python3-pysyncobj` to the build-tree packages (which wasn't the case so far). Anything else that is missing?

**Comment by CyberDem0n** [comment]
It is not related to pysyncobj. Missing validators were added in the https://github.com/zalando/patroni/commit/3a87d0e99b30e847ee313a071d7672b0087784ec. Could it be that port  2222 is busy?
What if you change it to something else in the test_validator.py?

**Comment by CyberDem0n** [comment]
Scratch it. Looks like `patroni.dcs.raft` could not be loaded.

**Comment by mbanck** [comment]
So anything I could/should do here?

**Comment by CyberDem0n** [comment]
Prerequisite: `python3-pysyncobj` should be installed.
After that please do the following:
```
$ python3
Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import patroni.dcs.raft
```

It could be that `patroni.dcs.raft` fails to load `pysyncobj` and you will see something like below:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/git/patroni/patroni/dcs/raft.py", line 9, in <module>
    from pysyncobj import SyncObj, SyncObjConf, replicated, FAIL_REASON
ModuleNotFoundError: No module named 'pysyncobj'
```

**Comment by mbanck** [comment]
I get
```
>>> import patroni.dcs.raft
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'patroni.dcs.raft'
```

**Comment by mbanck** [comment]
```
>>> import patroni.dcs
>>> modules = patroni.dcs.dcs_modules()
>>> for module_name in modules:          
...      print("%s", module_name)
... 
%s patroni.dcs.consul
%s patroni.dcs.etcd
%s patroni.dcs.etcd3
%s patroni.dcs.exhibitor
%s patroni.dcs.kubernetes
%s patroni.dcs.zookeeper
```

**Comment by CyberDem0n** [comment]
Do you run it while inside patroni repo?

**Comment by CyberDem0n** [comment]
Weird. Is there `patroni/dcs/raft.py` file?

**Comment by mbanck** [comment]
Yes, there is (`/usr/lib/python3/dist-packages/patroni/dcs/raft.py`), however, that file gets installed always anyway, no matter whether `pysyncobj` is present or not I believe. At least it is also in the current 2.0.1 Debian package, which does not have the `python3-pysyncobj` (Build-)dependency: 

```
$ dpkg -c patroni_2.0.1-3_all.deb  | grep raft.py
-rw-r--r-- root/root     16456 2020-10-01 15:18 ./usr/lib/python3/dist-packages/patroni/dcs/raft.py
$ dpkg -I patroni_2.0.1-3_all.deb  | grep sync
$
```

I won't rule out that something weird in the Debian packaging is causing this, but I've looked twice now and couldn't see anything obvious.

**Comment by CyberDem0n** [comment]
Patroni installed from pgdg on Ubuntu 18.04 shows the following (even without pysyncobj):
```python
$ python3
Python 3.6.9 (default, Oct  8 2020, 12:12:24) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import patroni.dcs
>>> patroni.dcs.dcs_modules()
['patroni.dcs.consul', 'patroni.dcs.etcd', 'patroni.dcs.etcd3', 'patroni.dcs.exhibitor', 'patroni.dcs.kubernetes', 'patroni.dcs.raft', 'patroni.dcs.zookeeper']
>>> import patroni.dcs.raft
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/patroni/dcs/raft.py", line 8, in <module>
    from pysyncobj import SyncObj, SyncObjConf, replicated, FAIL_REASON
ModuleNotFoundError: No module named 'pysyncobj'
```

So I don't have an explanation...

**Comment by mbanck** [comment]
Well from my point of view this looks like a regression from 2.0.1 (which is what the pgdg Ubuntu 18.04 should be at) to 2.0.2, but I'm having some troubles to rebuild/retest 2.0.1 right now.

I'll be looking further later on

**Comment by CyberDem0n** [comment]
@mbanck hmm, you remove raft files before running tests: https://salsa.debian.org/postgresql/patroni/-/blob/master/debian/tests/test

It might explain the behavior.

**Comment by mbanck** [comment]
d'oh, thanks!

**Comment by mbanck** [comment]
So I think I did remove those files earlier, because on all PGDG-supported releases (Debian buster/stretch, Ubuntu groovy/focal/bionic) there is no `pysyncobj`, so we can't add those Build-Depends.

Is there maybe some way to tell Patroni to skip the raft-based tests? Otherwise, I'll see about removing/patching out the new ones as well.

**Comment by CyberDem0n** [comment]
Something like following will skip raft tests if pysyncobj is not available:
```diff
diff --git a/tests/test_raft.py b/tests/test_raft.py
index 4b51ebe..bd61808 100644
--- a/tests/test_raft.py
+++ b/tests/test_raft.py
@@ -4,8 +4,14 @@ import tempfile
 import time
 
 from mock import Mock, patch
-from patroni.dcs.raft import DynMemberSyncObj, KVStoreTTL, Raft, SyncObjUtility
-from pysyncobj import SyncObjConf, FAIL_REASON
+
+try:
+    from patroni.dcs.raft import DynMemberSyncObj, KVStoreTTL, Raft, SyncObjUtility
+    from pysyncobj import SyncObjConf, FAIL_REASON
+except ImportError:
+    skip_tests = True
+else:
+    skip_tests = False
 
 
 def remove_files(prefix):
@@ -23,6 +29,7 @@ def remove_files(prefix):
                     time.sleep(1.0)
 
 
+@unittest.skipIf(skip_tests, 'pysyncobj is not available')
 @patch('pysyncobj.tcp_server.TcpServer.bind', Mock())
 class TestDynMemberSyncObj(unittest.TestCase):
 
@@ -57,6 +64,7 @@ class TestDynMemberSyncObj(unittest.TestCase):
         utility._onMessageReceived(0, '')
 
 
+@unittest.skipIf(skip_tests, 'pysyncobj is not available')
 class TestKVStoreTTL(unittest.TestCase):
 
     def setUp(self):
@@ -120,6 +128,7 @@ class TestKVStoreTTL(unittest.TestCase):
         so.destroy()
 
 
+@unittest.skipIf(skip_tests, 'pysyncobj is not available')
 class TestRaft(unittest.TestCase):
 
     _TMP = tempfile.gettempdir()
```

**Comment by mbanck** [comment]
This resulted in collection errors; not sure why also in `tests/test_raft.py`, but anyway also in `patroni/dcs/raft.py`:
```
collected 462 items / 6 errors / 456 selected                                                                                                                                                                                                

=================================================================================================================== ERRORS ===================================================================================================================
____________________________________________________________________________________________________ ERROR collecting tests/test_raft.py _____________________________________________________________________________________________________
tests/test_raft.py:34: in <module>
    class TestDynMemberSyncObj(unittest.TestCase):
tests/test_raft.py:41: in TestDynMemberSyncObj
    ???
E   NameError: name 'SyncObjUtility' is not defined
____________________________________________________________________________________________________ ERROR collecting tests/test_raft.py _____________________________________________________________________________________________________
tests/test_raft.py:34: in <module>
    class TestDynMemberSyncObj(unittest.TestCase):
tests/test_raft.py:41: in TestDynMemberSyncObj
    ???
E   NameError: name 'SyncObjUtility' is not defined
_______________________________________________________________________________________________ ERROR collecting tests/test_raft_controller.py _______________________________________________________________________________________________
tests/test_raft_controller.py:6: in <module>
    from pysyncobj import SyncObj
E   ModuleNotFoundError: No module named 'pysyncobj'
_______________________________________________________________________________________________ ERROR collecting tests/test_raft_controller.py _______________________________________________________________________________________________
ImportError while importing test module '/tmp/autopkgtest.szM8qK/build.bOh/src/tests/test_raft_controller.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_raft_controller.py:6: in <module>
    from pysyncobj import SyncObj
E   ModuleNotFoundError: No module named 'pysyncobj'
________________________________________________________________________________________________ ERROR collecting patroni/raft_controller.py _________________________________________________________________________________________________
patroni/raft_controller.py:5: in <module>
    from patroni.dcs.raft import KVStoreTTL
patroni/dcs/raft.py:9: in <module>
    from pysyncobj import SyncObj, SyncObjConf, replicated, FAIL_REASON
E   ModuleNotFoundError: No module named 'pysyncobj'
____________________________________________________________________________________________________ ERROR collecting patroni/dcs/raft.py ____________________________________________________________________________________________________
patroni/dcs/raft.py:9: in <module>
    from pysyncobj import SyncObj, SyncObjConf, replicated, FAIL_REASON
E   ModuleNotFoundError: No module named 'pysyncobj'
```

I had another look at the changes between 2.0.1 and 2.0.2 and reverting 3a87d0e99b does the trick (along with removing the raft-related files as before)

**Comment by CyberDem0n** [comment]
I see, it is probably more complex, like importing raft and pysyncobj directly from test methods.
I am not even sure that it would be worth the time spent on it.

If you have an issue with running tests, you can remove tests/test*raft*.py files instead of reverting the commit implementing config validation. The dcs/raft.py and patroni_raft_controller.py could stay. Since there will be no tests for them, tests will not fail.

**Comment by mbanck** [comment]
> If you have an issue with running tests, you can remove tests/test_raft_.py files instead of reverting the commit implementing config validation. The dcs/raft.py and patroni_raft_controller.py could stay. Since there will be no tests for them, tests will not fail.

I tried that, but I have the feeling (my knowledge of how the python unit tests work is still limited) that it tries to "collect" stuff at the beginning also from `patroni/dcs/raft.py` and fails to import `pysyncobj` there already, before running any actual tests.

I've uploaded 2.0.2 to Debian with the mentioned commit reverted and all raft-based files removed. I'll look into backporting `python3-pysyncobj`, either via Debian backports or PGDG in order to get raft functionality in, but it's not on the top of my todo list.
